### ISSUE TITLE
Is there a jq --slurp builtin .ie the inverse of .[]

### ISSUE THREAD

**Issue by themadsens** [stedolan/jq#2246]
After extensive googling, I can't find the inverse operation of .[] as a jq operator. Ie:

    echo '[1,2,3]' | jq '(.[] | <slurp>) == [1,2,3]' # expect true

This was also asked in #977 but there was no real answer given

**Comment by pkoppstein** [comment]
The inverse of `.[]` when applied to arrays is `[...]`, i.e., assuming the input is an array,  `[.[]]` evaluates to `.`, as does `[.][]`.

However, there is no inverse of `.[]` when applied to JSON objects as `.[]` loses all information about the keys.

**Comment by MrEndo13039** [comment]
The `[inputs]` does not satisfy the OP's question, which also seems to be the intent of #997 (also unsatisfied there, yet it was closed).
```
echo '[1,2,3]' | jq '(.[] | <slurp>) == [1,2,3]' # expect true
```

When substituting `[inputs]` filter for <slurp>, we get
```
echo '[1,2,3]' | jq '(.[] | [inputs]) == [1,2,3]' # expect true
```
and the result of that is:
```
[]
[]
[]
```
Alternately, with the "-n" option, 
`jq: error (at <unknown>): Cannot iterate over null (null)`

What he (and I) is looking for is a filter which will perform the job of "slurp" on its inputs to make an array.  It seems that `[inputs]` does not actually perform a slurp on its inputs, but rather "inputs" means something besides "the input to this specific filter." (I think it means "the same input as jq is using, like stdin", which would be already exhausted in this case.)

**Comment by itchyny** [comment]
As commented by @pkoppstein and also by https://github.com/stedolan/jq/issues/977#issuecomment-145378800, you can always build an array from a stream by wrapping `[` and `]`. So you can `echo '[1,2,3]' | jq '[.[]]'`. The array-like syntax `[1, 2, 3]` is one of them; it just builds an array from the stream `1, 2, 3` which emits 3 values. The query `[..]` builds an array from a stream which emits the json values recursively. The `inputs` is a stream emitting the renaming program inputs (the first value is consumed as the query input unless `-n` is used) and `[inputs]` build an array from them.

**Comment by MrEndo13039** [comment]
Thank you for the additional response, @itchyny .  I don't know what I did wrong initially.  I have had to do the workaround I proposed in the past because my "wrap in an array" technique just made a sequence of arrays, each one having a single input item; however, using `jq '[.[]]'` works fine for me today.  Perhaps it is because I just upgraded jq on my development system.  I will need to look for a more complex example in my code to see if it still runs incorrectly with the newer version.

**Comment by MrEndo13039** [comment]
My code still fails with new version of jq.  Here is a distilled example of what does not work as I expect it to:
```
cat > /tmp/test.json
{"id":1}
{"id":2}
{"id":3}
```

The slurp option, which is only available as a command line option, properly takes each item in the input and makes it an element in an array:
```
jq -s '.' /tmp/test.json 
[
  {
    "id": 1
  },
  {
    "id": 2
  },
  {
    "id": 3
  }
]
```
And now wrap the input in an array option, but that takes each element and puts it in its own array which is not the desired result:
```
jq '[.]' /tmp/test.json 
[
  {
    "id": 1
  }
]
[
  {
    "id": 2
  }
]
[
  {
    "id": 3
  }
]
```
So in this particular example problem, "slurp" is an potential solution simply because I put the distilled example into a file and that is the starting point for this example.  However, assume that there is an invocation of jq, which through a series of `select` and other manipulation ends up with an output which matches the test,json file, but it is the output in memory to be sent as input directly to the next filter.  This example demonstrates that one can't just wrap the whole thing in `[]` and have jq make an array with all of the elements of the input.
The following comment from #977 may be key here: "So if you want the output to have a single json array across all of the input items, you are going to need --slurp or similar."  That begs the question: if jq can do it with --slurp, then why can't jq do it inside an expression?
Is there another way, besides outputting to a file and starting a new "jq --slurp" instance?
@themadsens sorry if I incorrectly hijacked your thread; I feel that your question and #977 represent the same issue I am describing, but I could be wrong.

**Comment by itchyny** [comment]
You can always collect a value from a stream inside a query with `[ ... ]`. But the input value against a query is always single JSON value. jq repeatedly run the query against the jsons in stdin (or argument json file). So `jq '[.]'` just wraps each json values. The `.` query does nothing against the input and the query input is a single value, thus the output looks like that.
The slurp option (`-s`) changes the input to the jq program. It reads all the input values and build an array for the query input. Using with the raw input option (`-R`) means reading the entire input as a string.
The `inputs` are special stream that emits the remaining JSON values given to the jq program. Thus `jq -n '[inputs]'` is almost equivalent of `jq -s '.'`. The comment _--slurp or smiliar_ suggests the `inputs` function.

**Comment by bukzor** [comment]
I agree that `[.]` has the desired behavior, but the reason this isn't satisfactory is that it can't be used as a function. Consider:

```
seq 3 | jq --slurp 'def my_slurp: {{answer here}}; .[] | my_slurp'  # expect [1, 2, 3]
```

Now, if the answer is that `slurp` is indefinable in jq, even as a builtin, due to semantics / architecture of the language, that's an answer too, but `[.]` isn't the answer to this specific question.

**Comment by wtlangford** [comment]
tl;dr- it can't be defined as a function in jq.

jq's implementation uses a forking stack. Any time an expression does something that can produce multiple outputs, such as `.[]`, (and in some other places) the entire stack is forked and we mark a forkpoint. When the jq program hits a `RET` (return) opcode, it begins backtracking until it reaches a forkpoint and then continues forwards again from that forkpoint (but down the other path). (There are also other ways to backtrack, most notably `empty`, but also any expression that produces _no_ outputs).
The way `[ <someExpression> ]` works is by using what we call a `collect` internally. Essentially, it creates a fork points, instantiates an array, runs the expression, and then appends the output of that expression to the array, and forcibly backtracks. When it finally backtracks out of the expression to the original forkpoint, that fork continues with your jq program.

What this means, though, is that a `collect` expression _must_ wrap another expression.

**Comment by bukzor** [comment]
I suspect a way to rephrase this as a feature request is:

May we please have a functional / postfix syntax for the array constructor?

As an example, it would make sense (to me) to add a builtin named slurp, where x | y | slurp is equivalent to [x | y]. I suspect this would need to be implemented in the language parser, making it impossible as a user-defined function.

**Comment by bukzor** [comment]
I suspect a way to rephrase this as a feature request is:

> May we please have a functional / left-associative / postfix syntax for the array constructor?

As an example, it would make sense (to me) to add a builtin named `slurp`, where `x | y | slurp` is equivalent to `[x | y]`. I suspect this would need to be implemented in the language parser, making it impossible as a user-defined function.

If that change is impossible or infeasible or undesirable, we can finally close this issue as wontfix.

**Comment by themadsens** [comment]
Thanks @pkoppstein and @bukzor - That pretty much wraps it up.

**Comment by kpym** [comment]
@themadsens Is this issue closed as "wontfix" or for some other reason ?

**Comment by Pfiver** [comment]
And if it _is_ closed as "wontfix", is it "wontfix" because it is infeasible or undesirable?

**Comment by pkoppstein** [comment]
@Pfiver - wtlangford already explained:

> it can't be defined as a function in jq.

Note though that one can define functions with stream arguments; such functions can construct arrays. Consider the following (completely useless) example:

     jq -n 'def array(s): [s]; array(inputs)'

**Comment by Pfiver** [comment]
Never grasped the actual meaning of "can't" in software .... :-)

POC for "slurp": https://github.com/Pfiver/jq/commit/0ea8c17533317ec7676d5f7484162ecc94bbae29

```
➜  jq git:(master) ✗ echo '[1,2,3]' | ./jq '(.[] | slurp) == [1,2,3]' 
true

➜  jq git:(master) ✗ make check
...
============================================================================
Testsuite summary for jq 1.6-137-gd18b2d0-dirty
============================================================================
# TOTAL: 7
# PASS:  7
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
```

**Comment by yertto** [comment]
The [IO section in the manual](https://stedolan.github.io/jq/manual/#IO) explains how `inputs` works:
> `input`
> Outputs one new input.
>
> `inputs`
> Outputs all remaining inputs, one by one.
>
> This is primarily useful for reductions over a program's inputs.

... but perhaps it could do with some examples?
(see https://github.com/stedolan/jq/pull/2470)

**Comment by andrew-pickin-epi** [comment]
Just a note here (again) that using `inputs` missed the point of the OP. 
I think what was desired is a slurp function to works on the filtered results not `inputs` (otherwise one can just use `-s`)  eg.
 ```
 jq -r '. | select(.key=="yes") | slurp' << EOF
{"key":"yes", "value":5}
{"key":"no", "value":2}
{"key":"yes", "value":7}
EOF
```
```
[
  {
    "key": "yes",
    "value": 5
  },
  {
    "key": "yes",
    "value": 7
  }
]
```

Not the same as using `inputs`:
```
$ jq -r '. | select(.key=="yes") | [., inputs]' << EOF
{"key":"yes", "value":5}
{"key":"no", "value":2}
{"key":"yes", "value":7}
EOF
```
```
[
  {
    "key": "yes",
    "value": 5
  },
  {
    "key": "no",
    "value": 2
  },
  {
    "key": "yes",
    "value": 7
  }
]

```

**Comment by andrew-pickin-epi** [comment]
@Pfiver paybe your solution as a PR?

**Comment by yertto** [comment]
I still think some more examples of slurping using either `inputs` or the `--slurp` option would be helpful for folks trying to solve this _"inverse of .[]" issue_ in their own programs.

And I wonder if there's confusion as `inputs` is meant for slurping the _program's_ inputs, not inputs to a _filter_. 

@andrew-pickin-epi in your example its *still* possible to use `inputs`...

Here's some examples that will *all* output your desired array, by using either `inputs` or the `--slurp` option.
ie.
```json
[
  {
    "key": "yes",
    "value": 5
  },
  {
    "key": "yes",
    "value": 7
  }
]
```

#### 1. Using an extra `jq` command with the `--slurp` option.
```sh
jq 'select(.key=="yes")' << EOF | jq --slurp '.'
{"key":"yes", "value":5}
{"key":"no", "value":2}
{"key":"yes", "value":7}
EOF
```

#### 2. Using an extra `jq` command (but with `inputs` to slurp the _program's_ input, instead of the `--slurp` option).
```sh
jq 'select(.key=="yes")' << EOF | jq --null-input '[inputs]'
{"key":"yes", "value":5}
{"key":"no", "value":2}
{"key":"yes", "value":7}
EOF
```

#### 3. Using an extra `jq` command with `inputs` (but avoiding having to use the `--null-input` option).
```sh
jq 'select(.key=="yes")' << EOF | jq '[., inputs]'
{"key":"yes", "value":5}
{"key":"no", "value":2}
{"key":"yes", "value":7}
EOF
```

#### 4. Composing the two `jq` commands into a single command.
```sh
jq '[., inputs | select(.key=="yes")]' << EOF
{"key":"yes", "value":5}
{"key":"no", "value":2}
{"key":"yes", "value":7}
EOF
```

#### 5. Or  _after_ the `inputs` have all been _slurped_ , this example will `map` over the slurped array.
```sh
jq '[., inputs] | map(select(.key=="yes"))' << EOF
{"key":"yes", "value":5}
{"key":"no", "value":2}
{"key":"yes", "value":7}
EOF
```

#### 6. Or avoid the somewhat confusing `., ` by bringing back the `--null-input` option.
```sh
jq --null-input '[inputs] | map(select(.key=="yes"))' << EOF
{"key":"yes", "value":5}
{"key":"no", "value":2}
{"key":"yes", "value":7}
EOF
```

#### 7. Or avoid the `map` by applying the filter to each of the inputs.
```sh
jq --null-input '[inputs | select(.key=="yes")]' << EOF
{"key":"yes", "value":5}
{"key":"no", "value":2}
{"key":"yes", "value":7}
EOF
```

#### 8. Or avoid `inputs` by bring back the `--slurp` option.
```sh
jq --slurp 'map(select(.key=="yes"))' << EOF
{"key":"yes", "value":5}
{"key":"no", "value":2}
{"key":"yes", "value":7}
EOF
```

#### X. And for completeness a naive/wrong solution (that we're trying to avoid by doing all this slurping) that doesn't produce the desired output.
```sh
jq '[select(.key=="yes")]' << EOF
{"key":"yes", "value":5}
{"key":"no", "value":2}
{"key":"yes", "value":7}
EOF
```
```json
[
  {
    "key": "yes",
    "value": 5
  }
]
[]
[
  {
    "key": "yes",
    "value": 7
  }
]
```

---

IMO `4.` is probably the best solution as it applies the filter to all of the inputs while slurping them into an array, but allows `jq` to be invoked without  the `--slurp` *or* `--null-input` options.

**Comment by iainelder** [comment]
My filters are usually a long chain of pipes that operate on a stream. I write it like that because it's easier to read.

```jq
.objects[]
| longFilterExpression1
| longFilterExpression2
| longFilterExpression3
```

That's great until I need to `group_by` or `sort` at some point in the stream. Unfortunately I can't feed a stream to `sort` and get back a sorted stream, so I need to convert the stream to an array. That's where I would use the `slurp` operator.

```jq
.objects[]
| longFilterExpression1
| longFilterExpression2
| longFilterExpression3
| slurp | sort | .[]
| longFilterExpression4
```

I don't want to wrap some earlier part in an array and then immediately unwrap the array because it's ugly. But jq syntax currently doesn't offer a better solution.

```jq
[
  .objects[]
  | longFilterExpression1
  | longFilterExpression2
  | longFilterExpression3
] | sort | .[]
| longFilterExpression4
```
